archive/demos/shiny-drum-machine.js

```diff
@@ -6,396 +10,171 @@ window.AudioContext = window.AudioContext || window.webkitAudioContext;
+    modal: {
+        save: null,
+        load: null,
+    },
+});
+
+class Kit {
+    constructor(id, prettyName) {
+        this.id = id;
+        this.prettyName = prettyName;
+        this.buffer = {};
+    }
 
-Kit.prototype.load = function() {
-    if (this.startedLoading)
-        return;
-        
-    this.startedLoading = true;
-        
-    var pathName = this.pathName();
-
-    var kickPath = pathName + "kick.wav";
-    var snarePath = pathName + "snare.wav";
-    var hihatPath = pathName + "hihat.wav";
-    var tom1Path = pathName + "tom1.wav";
-    var tom2Path = pathName + "tom2.wav";
-    var tom3Path = pathName + "tom3.wav";
-
-    this.loadSample(0, kickPath, false);
-    this.loadSample(1, snarePath, false);
-    this.loadSample(2, hihatPath, true);  // we're panning only the hihat
-    this.loadSample(3, tom1Path, false);
-    this.loadSample(4, tom2Path, false);
-    this.loadSample(5, tom3Path, false);
-}
+    getSampleUrl(instrumentName) {
+        return `sounds/drum-samples/${this.id}/${instrumentName.toLowerCase()}.wav`;
+    }
 
-Kit.prototype.loadSample = function(sampleID, url, mixToMono) {
-    // Load asynchronously
-
-    var request = new XMLHttpRequest();
-    request.open("GET", url, true);
-    request.responseType = "arraybuffer";
-
-    var kit = this;
-
-    request.onload = function() {
-        context.decodeAudioData(
-            request.response,
-            function(buffer) {
-                switch (sampleID) {
-                    case 0: kit.kickBuffer = buffer; break;
-                    case 1: kit.snareBuffer = buffer; break;
-                    case 2: kit.hihatBuffer = buffer; break;
-                    case 3: kit.tom1 = buffer; break;
-                    case 4: kit.tom2 = buffer; break;
-                    case 5: kit.tom3 = buffer; break;
-                }
-
-                kit.instrumentLoadCount++;
-                if (kit.instrumentLoadCount == kit.instrumentCount) {
-                    kit.isLoaded = true;
-
-                    if (kit.demoIndex != -1) {
-                        beatDemo[kit.demoIndex].setKitLoaded();
-                    }
-                }
-            },
-            
-            function(buffer) {
-                console.log("Error decoding drum samples!");
-            }
-        );
+    load() {
+        const instrumentPromises = INSTRUMENTS.map(instrument => this.loadSample(instrument.name));
+        const promise = Promise.all(instrumentPromises).then(() => null);
+        // Return original Promise on subsequent load calls to avoid duplicate loads.
+        this.load = () => promise;
+        return promise;
     }
 
-    request.send();
+    async loadSample(instrumentName) {
+        const request = new Request(this.getSampleUrl(instrumentName));
+        const response = await fetch(request);
+        const responseBuffer = await response.arrayBuffer();
+
+        // TODO: Migrate to Promise-syntax once Safari supports it.
```

Comment:
Ditto on TODO. (i.e. issue tracker)

Can you elaborate on Safari problem?

Reply:
Seee https://developer.mozilla.org/en-US/docs/Web/API/BaseAudioContext/decodeAudioData#browser_compatibility

Reply:
Thanks!